lib/hanami/slice.rb

```diff
@@ -115,12 +115,20 @@ def define_container
 
           if root.join("lib").directory?
             config.component_dirs.add "lib" do |dir|
-              dir.default_namespace = namespace_path.tr(File::SEPARATOR, config.namespace_separator)
+              if application.configuration.autoloader
+                # If we're using an autoloader, expect all files in the slice's `lib/` to
+                # use the slice's const namespace
+                dir.namespaces.root const: namespace_path.tr(File::SEPARATOR, config.namespace_separator)
+              else
+                # Otherwise expect them to be inside `lib/[namespace]`, and make that take
+                # the root identifier namespace in the container
+                dir.namespaces.add namespace_path
+              end
```

Comment:
If we're not autoloading, then I think it's reasonable to expect users to have their slices set up so the file system path matches the nested constant (i.e. `lib/foo/blah.rb` defining `Foo::Blah`).

This is our escape hatch that still allows the system to behave properly even if the autoloader is disabled.

What I wonder, though, is whether we need to offer any further configurability than this? I feel like we probably don't, at least not right away, since it would result in a whole lot of complexity and would probably require us to build some higher-level configuration abstraction over a bunch of these dry-system settings.

I think what would probably be better is just making sure the idea we have of "bring your own container" works as intended. Right now it's not really something out tests exercise.